core/dispatcher/src/main/scala/whisk/core/dispatcher/Dispatcher.scala

```diff
@@ -102,57 +87,42 @@ trait MessageDispatcher extends Registrar with Logging {
+     * For every message that is received, this method extracts the path property
      * from the message and checks if there are registered handlers for the message.
      * A handler is registered via addHandler and unregistered via removeHandler.
      * All matches are checked in parallel, and messages are dispatched to all matching
      * handlers. The handling of a message is wrapped in a Future. A handler is skipped
      * if it is not active.
      */
-    def start() = if (!started) {
-        consumer.onMessage((topic, bytes) => {
-            val raw = new String(bytes, "utf-8")
-            val msg = Message(raw)
-            msg match {
-                case Success(m) =>
-                    implicit val tid = m.transid
-                    if (m.path.nonEmpty) {
-                        inform(handlers.par) foreach {
-                            case (name, handler) =>
-                                val matches = handler.matches(m.path)
-                                handleMessage(handler, topic, m, matches)
-                        }
-                    }
-                    true
-                case Failure(t) =>
-                    info(this, errorMsg(raw, t))
-                    true
-            }
-        })
-        started = true
-    }
-
-    /** Stops the message dispatch and closes the producer and consumer. */
-    def stop() = if (started) {
-        consumer.close()
-        started = false
+    def process(topic: String, bytes: Array[Byte]) = {
+        val raw = new String(bytes, "utf-8")
+        Message(raw) match {
+            case Success(m) =>
+                implicit val tid = m.transid
+                if (m.path.nonEmpty) inform(handlers) foreach {
+                    case (name, handler) =>
+                        val matches = handler.matches(m.path)
+                        handleMessage(handler, topic, m, matches)
+                }
+            case Failure(t) => info(this, errorMsg(raw, t))
+        }
     }
 
     private def handleMessage(rule: DispatchRule, topic: String, msg: Message, matches: Seq[Match]) = {
-        implicit val ec = executionContext
-
         implicit val tid = msg.transid
+        implicit val executionContext = actorSystem.dispatcher
+
         if (matches.nonEmpty) Future {
             val count = counter.next()
             info(this, s"activeCount = $count while handling ${rule.name}")
-            rule.doit(topic, msg, matches)
+            rule.doit(topic, msg, matches) // returns a future which is flat-mapped to hang onComplete
         } flatMap (identity) onComplete {
             case Success(a) => info(this, s"activeCount = ${counter.prev()} after handling $rule")
             case Failure(t) => error(this, s"activeCount = ${counter.prev()} ${errorMsg(rule, t)}")
         }
     }
 
-    private def inform(matchers: ParTrieMap[String, DispatchRule])(implicit transid: TransactionId) = {
```

Comment:
remove Par collection since there is only one invoker (and generally one consumer per dispatcher).
it's arguable if we should flatten this further - i opted to defer for now.